### PR TITLE
[3.3] Remove remaining Twig global use

### PR DESCRIPTION
**Pull request by GawainLynch**
As per discussion in Tuesday's meeting.

@CarsonF On the `BackendBase` controller one, I know, I know … but scope … believe me they are used all over backend templates, and refactoring the entire backend template logic this close to release is just not on 

… my O.C.D. too :disappointed:

### CODE REVIEW

File: app/view/twig/_base/_listing.twig [hunk 1]
```diff
@@ -301,13 +301,13 @@ row_header:     prop.extended and (prop.first or (prop.has_groupname and prop.ne
                     {% for taxonomyslug, values in content.taxonomy %}
                         {% if values|length > 1 %}
                             <li>
-                                <a class="nolink">{{ config.get('taxonomy')[taxonomyslug].name }}:
+                                <a class="nolink">{{ app.config.get('taxonomy')[taxonomyslug].name }}:
```
Comment (CarsonF) [inline app/view/twig/_base/_listing.twig]: Are we deprecating the `config` global?

File: app/view/twig/_nav/_footer.twig [hunk 2]
```diff
@@ -3,15 +3,15 @@
     <footer id="bolt-footer" class="hidden-xs {% if global.request.cookies.get('sidebar') %}bolt-footer-hidden{% endif %}">
         {{ __('general.phrase.provided-by-colon') }}
         {{ app.config.get('general/branding/provided_link')|raw }} &ndash;
-        <i class="fa fa-cog"></i> <b>Bolt {{ bolt_version }}</b>:
+        <i class="fa fa-cog"></i> <b>Bolt {{ context.bolt_version }}</b>:
```
Comment (CarsonF) [inline app/view/twig/_nav/_footer.twig]: The constant can be used here instead.

File: src/Controller/Backend/BackendBase.php [hunk 3]
```diff
@@ -42,6 +42,14 @@ protected function render($template, array $context = [], array $globals = [])
             $context = ['context' => $context];
         }
 
+        /** @deprecated Deprecated since 3.0, to be removed in 4.0. */
+        $oldGlobals = $this->app['twig']->getGlobals();
+        foreach (['user', 'users', 'bolt_version'] as $key) {
+            if (!isset($context['context'][$key])) {
+                $context['context'][$key] = $oldGlobals[$key];
+            }
+        }
```
Comment (CarsonF) [inline src/Controller/Backend/BackendBase.php]: Why? These are already globals that can be lazily eval'ed.

File: theme/base-2016/extrafields.twig [hunk 4]
```diff
@@ -4,7 +4,7 @@
 
     <h1>{{ record.title }}</h1>
 
-    {{ fields(template = 'partials/_sub_fields.twig') }}
+    {{ fields(record = record, template = 'partials/_sub_fields.twig') }}
```
Comment (CarsonF) [inline theme/base-2016/extrafields.twig]: This still doesn't handle the case where record is left out.

Comment (bobdenotter) [inline theme/base-2016/extrafields.twig]: In what scenario is `record` not present for a single record page?

Comment (CarsonF) [inline theme/base-2016/extrafields.twig]: I mean when the user doesn't pass record. i.e. The PHP code still tries to pull it from globals.

Comment (GawainLynch) [inline theme/base-2016/extrafields.twig]: > I mean when the user doesn't pass record. i.e. The PHP code still tries to pull it from globals.

Yes, but that is going to be gone in master … the point I wanted to make by setting that parameter is that it is going to be required in v4, start using it now. :wink:

File: app/view/twig/_base/_listing.twig (hunk 1 above)
Comment (GawainLynch) [inline app/view/twig/_base/_listing.twig]: You had it marked as so, and the template base was split between them … what's it going to be?

File: theme/base-2016/extrafields.twig (hunk 4 above)
Comment (CarsonF) [inline theme/base-2016/extrafields.twig]: Ok so we should add a deprecation message to that function then. And throw exception if twig_globals is disabled and a record isn't passed in.

File: src/Controller/Backend/BackendBase.php (hunk 3 above)
Comment (GawainLynch) [inline src/Controller/Backend/BackendBase.php]: If you set globals compatibility off, they fail … I'm all ears.

File: app/view/twig/_nav/_footer.twig (hunk 2 above)
Comment (GawainLynch) [inline app/view/twig/_nav/_footer.twig]: ¿que?

File: theme/base-2016/extrafields.twig (hunk 4 above)
Comment (GawainLynch) [inline theme/base-2016/extrafields.twig]: Starting a new branch now :+1:

File: app/view/twig/_nav/_footer.twig (hunk 2 above)
Comment (CarsonF) [inline app/view/twig/_nav/_footer.twig]:
```twig
{{ constant('Bolt\\Version::VERSION') }}
```

File: src/Controller/Backend/BackendBase.php (hunk 3 above)
Comment (CarsonF) [inline src/Controller/Backend/BackendBase.php]: https://github.com/bolt/bolt/blob/release/3.3/src/EventListener/ConfigListener.php#L158-L159

File: app/view/twig/_base/_listing.twig (hunk 1 above)
Comment (CarsonF) [inline app/view/twig/_base/_listing.twig]: Where did I have it marked as deprecated?

File: src/Controller/Backend/BackendBase.php (hunk 3 above)
Comment (GawainLynch) [inline src/Controller/Backend/BackendBase.php]: So `user` & `users` will remain as globals?

File: app/view/twig/_nav/_footer.twig (hunk 2 above)
Comment (GawainLynch) [inline app/view/twig/_nav/_footer.twig]: Done

File: theme/base-2016/extrafields.twig (hunk 4 above)
Comment (GawainLynch) [inline theme/base-2016/extrafields.twig]: Done

File: src/Controller/Backend/BackendBase.php (hunk 3 above)
Comment (CarsonF) [inline src/Controller/Backend/BackendBase.php]: Idk. Where are they used? Maybe not `users`, but I could see `user` being useful. Maybe a function would be better though that way access can be checked when requested?

Comment (CarsonF) [inline src/Controller/Backend/BackendBase.php]: If we don't want them as globals, and we want to pass them for all backend routes (which I may have a problem with), then they should be pulled the proper way, not from Twig globals.

File: app/view/twig/_base/_listing.twig (hunk 1 above)
Comment (GawainLynch) [inline app/view/twig/_base/_listing.twig]: That was my reading/remembering of https://github.com/bolt/bolt/blob/release/3.3/src/Twig/Extension/BoltExtension.php#L88-L90

File: src/Controller/Backend/BackendBase.php (hunk 3 above)
Comment (GawainLynch) [inline src/Controller/Backend/BackendBase.php]: Yeah, fine … what I am trying to get here is stuff in place so I can cut master over with as little friction between now and when 3.x EoLs … 

As I mention in the PR text, there is scope on this … I am **not** going to refactor the entire backend templates this close to release … what is the middle ground here? :smile:

File: app/view/twig/_base/_listing.twig (hunk 1 above)
Comment (GawainLynch) [inline app/view/twig/_base/_listing.twig]: If we're keeping config global, I am happy enough with that, but please make a call, following your decision tree isn't always easy over the wire :rofl:

File: src/Controller/Backend/BackendBase.php (hunk 3 above)
Comment (GawainLynch) [inline src/Controller/Backend/BackendBase.php]: > Where are they used

Oh, and as per original PR text, _all over the place_ … hence the problem needing a solution

Comment (CarsonF) [inline src/Controller/Backend/BackendBase.php]: > As I mention in the PR text, there is scope on this … I am not going to refactor the entire backend templates this close to release … what is the middle ground here? 😄

The leave them as globals, they work fine that way regardless of twig_globals flag.

Comment (GawainLynch) [inline src/Controller/Backend/BackendBase.php]: Done :+1:

**Pull request merged**